src/group_theory/iwasawa.lean

```diff
@@ -0,0 +1,483 @@
+if it is transitive, and if all stabilizers G_x are maximal subgroups.
+(The class `subgroup.is_maximal` organizes the notion of maximal subgroups.)
+
+**Lemma (`is_primitive_of_two_trans`).**
+If the action is 2-transitive, then it is primitive.
+
+(TODO : define n-transitive ations (for n:ℕ) and proves the inductive results.)
+
+**Theorem (`Iwasawa_Criterion.Iwasawa_mk`, Iwasawa criterion).**
+Let G be a group with a primitive action on a set X.
+Assume given an Iwasawa structure, namely,
+for all x:X, an abelian subgroup T x of G such that
+1. ∀ g:G x:X, T (g x) = g (T x) g⁻¹
+2. The union of all T x generates G
+Then every normal subgroup of G that acts nontrivially on X contains
+the derived subgroup D(G) of G.
+
+These hypotheses are bundled in the structure `has_iwasawa_structure`.
+
+**Corollary (`Iwasawa_Criterion.is_simple`).**
+Let G be a perfect group (`commutator G = ⊤`)
+with a faithful and primitive action on a set X that admits an Iwasawa structure.
+Then G is simple.
+
+## Simplicity of special linear groups
+
+**Theorem (TODO).** Let K be a field and let n be an integer such that n ≥ 2.
+Unless n = 2 and K has ≤ 3 elements, the group PSL(n,K) is simple.
+
+
+## Simplicity of alternate groups
+
+**Theorem (TODO).** Let n be an integer such that n ≥ 5.
+The alternate group A(n) is simple.
+
+-/
+
+
+
+section Maximal_Subgroups
+variables {G : Type*} [group G]
+
+namespace subgroup
+/-- An subgroup is maximal if it is maximal in the collection of proper subgroups. -/
+class is_maximal (K : subgroup G) : Prop :=
+(out : is_coatom K)
+
+theorem is_maximal_def {K : subgroup G} : K.is_maximal ↔ is_coatom K := ⟨λ h, h.1, λ h, ⟨h⟩⟩
+
+theorem is_maximal.ne_top {K : subgroup G} (h : K.is_maximal) : K ≠ ⊤ := (is_maximal_def.1 h).1
+
+theorem is_maximal_iff {K: subgroup G} : K.is_maximal ↔
+  K ≠ ⊤ ∧ ∀ (H: subgroup G) g, K ≤ H → g ∉ K → g ∈ H → H = ⊤ :=
+begin
+  split,
+  { intro hK, split, exact is_maximal.ne_top hK,
+  intros H g hKH hgK hgH,
+  suffices : K < H,
+    apply (is_maximal_def.1 hK).2, assumption,
+  have zKH : K ≠ H, { rw ne.def, intro z, rw z at hgK, exact hgK hgH },
+  exact (ne.le_iff_lt zKH).mp hKH,},
+  { rintros ⟨ hG, hmax ⟩,
+  split, split, assumption,
+  introsI H hKH,
+  obtain ⟨ g, hgH, hgK ⟩ := (set.exists_of_ssubset hKH),
+  exact hmax H g (le_of_lt hKH) hgK hgH, },
+end
+
+theorem is_maximal.eq_of_le {K H: subgroup G}
+  (hK : K.is_maximal) (hH : H ≠ ⊤) (KH : K ≤ H) : K = H :=
+eq_iff_le_not_lt.2 ⟨KH, λ h, hH (hK.1.2 _ h)⟩
+
+end subgroup
+end Maximal_Subgroups
+
+section Basic_Lemmas
+/-- If G and H have multiplications *
+and φ : G → H is a surjective multiplicative map,
+and if G is commutative, then H is commutative.
+
+Since I only use it with groups,
+I should probably use function.surjective.comm_semigroup
+--/
+lemma surj_to_comm {G H : Type*} [has_mul G] [has_mul H] (φ: mul_hom G H)
+   (is_surj : function.surjective φ) (is_comm : is_commutative G (*)) :
+    is_commutative H (*) :=
```

Comment:
mathlib indent style, and a better name:
```suggestion
lemma function.surjective.mul_commutative {G H : Type*} [has_mul G] [has_mul H] (φ : mul_hom G H)
  (is_surj : function.surjective φ) (is_comm : is_commutative G (*)) :
  is_commutative H (*) :=
```

Reply:
Also, I recommend using `commutative ((*) → G → G)` instead of `is_commutative G (*)`.

Reply:
Do you mean : `commutative ((*) : G → G → G)` ?

Reply:
(In any case, that worked as you indicated with this modification…)

Reply:
Yes I did, whoops!

Reply:
I overlooked other implications of this modification, for the moment, the file does not compile.

Reply:
There are a lot of other `is_commutative` to rewrite and I wonder whether it is not clearer to refer to the structural properties of objects (groups, etc.) rather than going back to their internal law.

Is there a way to shorten : `commutative ((*) : (quotient_group.quotient N) → (quotient_group.quotient N) → (quotient_group.quotient N))` ?

Reply:
Does `commutative ((*) : _ → _ → quotient_group.quotient N)` work?

Reply:
No, but you certainly know how to recover, if G is a group/monoid/... its `mul` law, or any of its types that are part of its structure. (What if there are many of them?)

Reply:
I don't understand the question

Reply:
I had hoped that something like `(quotient_group.quotient N).mul` would give me the `mul` part of the group (monoid) structure of `quotient_group.quotient N`. But that didn't work like that.